### PR TITLE
Added concurrency support

### PR DESCRIPTION
**Pull request by devconcept**
Sorry for the delay. Here is a new PR. I've been short of time lately. The algorithm is working with shorter, better syntax than the last PR. Everything is commented so unfamiliar code is easy to understand.

I considered using Observables (or to be precise Subjects) but I discarded the idea quickly because it probably produce more code with the exact same results, an extra dependency, and it didn't fit with the mental model of a request queue being drained.

As for use cases of concurrency; marketplace integrations are the likely candidate for this issue since you will have several clients (aka vendors) operating at the same time (even in a single transaction) and people will be tempted to combine calls when the source or the destination of the operation are different vendors. Despite `Promise.all` being a "fail-fast" mechanism you could choose to log and recover manually, feeding `Promise.all` with an iterable of promise chains that already handles logging the error for each failure somewhere in between. The fail handler of `Promise.all` is executed as soon as the first error is encountered but each individual chain always runs to completion unless they fail by themselves (if they don't fall into the error channel there is no reason to stop).

Also your data listing operations are great but they can't have notion of every developer business logic. Let's say I'm listing recent payments made from one vendor to other vendors. [Paging and filtering](https://developer.paypal.com/docs/api/overview/#query-parameters) falls short here because "vendor" is a business logic thing that is known only to my application. You could improve this feature be accepting a list of ids to conform a select query but in the meantime I can combine some request in a paginated view (lets say five items). I assure you there will be a `Promise.all` somewhere in the source code. This way I can make it work and fullfill any requirement even when the API lacks a certain feature. Different applications will have different business logic and is a fact that you won't be able to cover all scenarios. Payouts also lacks a listing feature which makes them another ideal candidate.

This are just a couple of examples. With other promise implementations the `all` method will not be the only method in question. I only mention it because is the most popular and commonly used. Its also a valid observation that the concurrency issue is a weird error that only happens from time to time, increasing in probability with traffic, but the purpose is to leave users with errors they can handle by themselves and expose only reliable behavior. I knew about this problem by looking at the source code so I knew that all the calls must be chained. How many people will do the same?

Other cases solved here were token caching, which was missing from the first draft of the `2.0-beta`. You can't fetch a token for every request, only when you don't have one or is expired. The caching is made in two levels because you could have the same client in different environments. It's up to you to decide if such a precaution is necessary. I only handled the worst case scenario here where two environments where used at the same time.

Remember that OpenId Connect features also have a `AccessTokenRequest` object but in my opinion a distinction should be made so people know they are completely different stuff.

Finally I added a constant to loose the token expiration computation because a human being is capable to realize that a token that is, for example, 100ms to expire will expire on transit and the call will fail anyway. The automatic recovery jumps in and solves the problem but you can save one round trip by slighly adjusting the calculation refreshing calls that are too close to expire. Only testing will tell you what "close to expire" means.

I hope this feedback is useful to you.

### CODE REVIEW

**Comment by sdcoffey** [comment]
Hey @devconcept,

Thanks so much for taking the time to put this together, honestly we really appreciate it. We're going to review this today. At first blush there might be some minor nomenclature to adjust, some tests to add, but this looks pretty solid to me.

**Comment by sdcoffey** [comment]
@devconcept,

Would you rather us make changes directly to this PR, or start a review with desired changes?

**Comment by devconcept** [comment]
Make changes directly. I can fetch the commits and add more if necessary.

**Comment by sdcoffey** [comment]
Just pushed a few commits here. You'll notice that we changed some names and simplified some codepaths, but the core of the work is unchanged. Give us a quick review here and we'll merge!

**Comment by devconcept** [comment]
Hi @sdcoffey

I found something that bothers me on the update. In the cache you have this code

```javascript
  static cacheForEnvironment(environment, refreshToken) {
    let key = environment.clientId;
    if (refreshToken) {
      key += `_${refreshToken}`;
    }
    /* ...... rest of the code */
  }
```

When there is a refresh token you create a new cache which doesn't make sense to me. The way I see it there is a one to one correspondence between a client and a token. I don't know how far PayPal has diverted from the spec, but a token is an object structure that may or may not contain a refresh token<sup>1</sup>. Those two are different things because a token has limited life, multiple uses and grants you access to the resources while refresh tokens have longer duration (many people make it unlimited) and only serves to obtain a new access token<sup>2</sup> (you can't access protected resources with them).

Having or not a refresh token influences how you obtain a new token. If you didn't receive a refresh token you just make a new request `grant_type=client_credentials` like if you never had a token. If you do however you get a new one by using `grant_type=refresh_token&refresh_token=${refreshToken}`. The new token replaces the old one and, if a new refresh token was issued in the token request, the current refresh token is replaced with the new<sup>3</sup>. This is why the token structure is entirely replaced on every update. On my implementation I included this possibility because it is the faithful spec implementation. OpenId Connect inherits from OAuth2 so the basic principles are still the same. Even when the current Client Credentials OAuth2 server does not implement refresh tokens right now, PayPal could add them in the future and your library would still "just work" with the update. Also OpenId Connect tokens could reuse the `AccessToken` class without modification (at least in the common denominator area between specs).

Also you removed the override to the execute method which is an integral part of the recovery mechanism. You cannot assume that you won't get a 401 error because you will if:

- Someone detects that a client has been compromised the best course of action could be revoking access token.

- No matter how relaxed the token expiration computation is there will be timing issues. This a performance optimization and not an insurance versus errors.

- An error on the app server clock could screw the token expiration computation. With the recovery mechanism you just set the clock again and the app never fails.

Catching 401 errors is still a necessity because you always update access before and after executing http requests. You try to prevent the error and recover if it happens. Remember that token expiration is a *hint*, never a *rule*.

<sup>1</sup> https://tools.ietf.org/html/rfc6749#section-1.5

> Issuing a refresh token is optional at the discretion of the authorization server. If the authorization server issues a refresh token, it is included when issuing an access token

<sup>2</sup> https://tools.ietf.org/html/rfc6749#section-1.5

> Unlike access tokens, refresh tokens are intended for use only with authorization servers and are never sent to resource servers.

<sup>3</sup> Sending a new refresh token is almost a standard because, if they were compromised, it would be very difficult to detect and protect. An attacker could continuously be obtaining access to protected resources by getting new access tokens. Refresh tokens usually have one time use and are discarded after that.

**Comment by sdcoffey** [comment]
Hey @devconcept,

I'm glad you brought up that point about refresh tokens here, it seems like there might be some confusion. The refresh token we're talking about here is in reference to API actions taken on a user's behalf. In this flow, you would use a client-side sdk to get an authorization code, which is then exchanged server-side for a refresh token as we're referring to it here. We allow you the option to pass in the refresh token to a `PayPalHttpClient` since we can handle getting access tokens that are scoped to that user under the hood. Check out [the PayPal docs](https://developer.paypal.com/docs/integration/mobile/make-future-payment/?mark=refresh%20token#exchange-the-auth-code-for-refresh-and-access-tokens) on that if that's not entirely clear.

You're absolutely right about catching authorization errors. However, it seemed to us that the way it was written before was to catch the error inside the HttpClient, and try to refresh the token again. If, for example, someone input their cleint id/secret credentials incorrectly, or as you say, there was some other problem with the authorization, this mechanism would automatically retry the authorization request endlessly, with no feedback to the developer. We feel that the developer should receive an error in these cases. In any case, this feels a bit like a premature optimization, and something we can add later if we get feedback and requests for it. Does that sound reasonable?

**Comment by devconcept** [comment]
Hi @sdcoffey

The reason why that code seemed odd to me was because it is not reflected in the old api in any way. Here are the important parts

```javascript
// ....
// line 12

var token_persist = {};

// ....
// line 71

client.invoke('POST', '/v1/oauth2/token', payload, http_options, function (err, res) {
  var token = null;
  if (res) {
    if (!config.authorization_code && !config.refresh_token) {
      var seconds = new Date().getTime() / 1000;
      token_persist[config.client_id] = res;
      token_persist[config.client_id].created_at = seconds;
    }

    if (!config.authorization_code) {
      token = res.token_type + ' ' + res.access_token;
    } else {
      token = res.refresh_token;
    }
  }
  cb(err, token);
});

// ....
// line 146

if (http_options.client_id in token_persist && !utils.checkExpiredToken(token_persist[http_options.client_id]) && !http_options.refresh_token) {
    http_options.headers.Authorization = "Bearer " + token_persist[http_options.client_id].access_token;
```

If you follow that trail you will see that caching is never used when refresh tokens are involved. My previous message was a little too theoretical (maybe this is the reason of the misunderstanding) but I wrote it to let you know the motivation and reasoning behind some parts of my implementation. Let me rephrase my previous comment.

I can see now that you are creating a cache to get your hands on a queue for refresh tokens too. The article you reference is from the Mobile SDK and it has a fundamental difference with the traditional Restful SDK operations. On the regular api operations you get a token and perform api calls, no additional steps are required. The mobile counterpart requires **user consent** first, which is an entirely different animal. If I just restart my server, all cache information is lost (and if I use a load balancer I'll get some really weird results). This is not an issue for the Rest SDK code which will get a new token and will continue working as expected. The mobile apps will crash because now you need to ask for user consent again (which is a big no-no for mobile apps). This is why this cache can only be used for a queue and not to persist tokens which is not straightforward for me to see because I understand the consequences.

What we are having here is an "orange and apples" conversation. What I was talking back there was the possibility for the Client Credentials flow to get refresh tokens as well as the Authorization Code flow does. I'm not a PayPal developer so I'm not aware of any plans/roadmaps/issues that this possibility will present. In that moment it seemed a no-brainer to add it. I apologize for removing refresh tokens from the constructor. That was a mistake from my end and you got it right from the beginning. Not seeing an authorization code anywhere was probably the reason for this confusion (how else will you get a refresh token?). I guess you are still debating on how to construct that request.

About the authorization errors, is easy to prove that the implementation is right. This is the code in question:

```javascript
_wait(request) {
  return this.client.wait(request).then(() => {
    this._setAuthHeader(request);
    return super.execute(request);
  });
}
```

If you look closely the request is retried with a call to `super.execute` which has no `catch` handler attached so if someone input their credentials wrong the call is retried once and then it bails out as expected. There will never be an infinite loop with this code. You can test it all you want. I also forgot to mention on my last message about [Dayligth Saving Time](https://en.wikipedia.org/wiki/Daylight_saving_time) which means now we know for sure that at least once a year every single application written with your api will crash simultaneously. This is already implemented in the old api and I showed you the consequences of not doing so in [this](https://github.com/paypal/PayPal-node-SDK/issues/234#issuecomment-332513150) message so, in my opinion, this is a mandatory feature and not a premature optimization. There is no point in waiting for the code to fail to fix a problem you already know is there.

**Comment by sdcoffey** [comment]
@devconcept,

Great, glad we're on the same page about the refresh token business. I think we can all agree that the terms are a little confusing. 

Re: retries, I hadn't considered the Daylight savings edge case. You were right about the logical flow of the code you had written. I've put back a version of the retry logic and added a test to validate that the auth token call is only made one additional time. Let me know what you think and we can finally get this thing merged!

**Comment by devconcept** [comment]
@sdcoffey,

Great. I just added one missing test for 401 errors, fix a small bug where you fetched the refresh token and returned the token object instead of the contents of the retried http request and remove all the unandled promise rejection warnings. Check the code but the contents are almost identical.

**Comment by sdcoffey** [comment]
@devconcept,

Awesome, looks great. I'm going to go ahead and merge this. Again, I can't thank you enough for all your help making the new version of this SDK even better. Please let us know if you find any other issues, or if you have any other feedback on our rewrite.
